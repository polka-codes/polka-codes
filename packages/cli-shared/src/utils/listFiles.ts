import { promises as fs } from 'node:fs'
import { join, relative, resolve } from 'node:path'
import ignore, { type Ignore } from 'ignore'

/** Default patterns commonly ignored in projects of various languages. */
const DEFAULT_IGNORES = [
  '__pycache__',
  '.DS_Store',
  '.env',
  '.git',
  '.idea',
  '.svn',
  '.temp',
  '.vscode',
  'coverage',
  'dist',
  'node_modules',
  'out',
  'Thumbs.db',
]

/**
 * Reads a `.gitignore` file in `dirPath` (if it exists) and appends its lines
 * to the `basePatterns`. Returns a new array without mutating the original.
 */
async function extendPatterns(basePatterns: string[], dirPath: string): Promise<string[]> {
  try {
    const gitignorePath = join(dirPath, '.gitignore')
    const content = await fs.readFile(gitignorePath, 'utf8')
    const lines = content.split(/\r?\n/).filter(Boolean)
    return [...basePatterns, ...lines]
  } catch {
    // No .gitignore or unreadable
    return basePatterns
  }
}

/** Creates an `ignore` instance from the given patterns. */
function createIgnore(patterns: string[]): Ignore {
  return ignore().add(patterns)
}

/**
 * Lists files under `dirPath` in BFS order, respecting:
 *   - A default set of ignores
 *   - A root .gitignore under `cwd`
 *   - Any .gitignore files in child directories (merged as we go)
 *
 * Returns `[files, limitReached]`:
 *   - `files` is the array of file paths (relative to `cwd`)
 *   - `limitReached` is `true` if `maxCount` was hit, otherwise `false`
 *   - When truncated, adds markers like `path/to/dir/(files omitted)` for truncated directories
 */
export async function listFiles(
  dirPath: string,
  recursive: boolean,
  maxCount: number,
  cwd: string,
  excludeFiles?: string[],
): Promise<[string[], boolean]> {
  // Merge default ignores with root .gitignore and excludeFiles (if found)
  let rootPatterns = [...DEFAULT_IGNORES, ...(excludeFiles || [])]
  try {
    const rootGitignore = await fs.readFile(join(cwd, '.gitignore'), 'utf8')
    const lines = rootGitignore.split(/\r?\n/).filter(Boolean)
    rootPatterns = [...rootPatterns, ...lines]
  } catch {
    // No .gitignore at root or unreadable; ignore silently
  }

  // Final results (relative to `cwd`) and indicator if we reached the limit
  const results: string[] = []

  // Track directories we've seen to avoid duplicate "(files omitted)" markers
  const processedDirs = new Set<string>()

  // BFS queue
  // Each entry holds the directory path, patterns, and relative path
  const queue: Array<{ path: string; patterns: string[]; relPath: string }> = [
    {
      path: resolve(dirPath),
      patterns: rootPatterns,
      relPath: relative(cwd, resolve(dirPath)).replace(/\\/g, '/') || '.',
    },
  ]

  // Perform BFS until queue is empty or maxCount is reached
  while (queue.length > 0) {
    // biome-ignore lint/style/noNonNullAssertion: checked above
    const { path: currentPath, patterns: parentPatterns, relPath: currentRelPath } = queue.shift()!

    // Mark this directory as processed
    processedDirs.add(currentRelPath)

    // Merge parent's patterns with local .gitignore
    const mergedPatterns = await extendPatterns(parentPatterns, currentPath)
    const folderIg = createIgnore(mergedPatterns)

    const entries = await fs.readdir(currentPath, { withFileTypes: true })
    entries.sort((a, b) => a.name.localeCompare(b.name)) // Sort entries for consistent order

    for (const entry of entries) {
      const fullPath = join(currentPath, entry.name)
      // Convert full path to something relative to `cwd`
      const relPath = relative(cwd, fullPath).replace(/\\/g, '/')

      if (folderIg.ignores(relPath)) {
        continue // Skip ignored entries
      }

      if (entry.isDirectory()) {
        if (recursive) {
          queue.push({
            path: fullPath,
            patterns: mergedPatterns,
            relPath,
          })
        }
      } else {
        results.push(relPath)
        if (results.length >= maxCount) {
          // We've hit the limit, add "(files omitted)" markers for directories
          // still in the queue and the current directory if we haven't processed all its files

          // First, check if there are remaining files in the current directory
          const remainingEntries = entries.slice(entries.indexOf(entry) + 1)
          const hasRemainingFiles = remainingEntries.some(
            (e) => !e.isDirectory() && !folderIg.ignores(relative(cwd, join(currentPath, e.name)).replace(/\\/g, '/')),
          )

          if (hasRemainingFiles) {
            const marker = `${currentRelPath}/(files omitted)`
            results.push(marker)
          }

          // Then add markers for all directories still in the queue
          for (const queueItem of queue) {
            // Only add markers for directories we haven't processed yet
            if (!processedDirs.has(queueItem.relPath)) {
              const marker = `${queueItem.relPath}/(files omitted)`
              results.push(marker)
              processedDirs.add(queueItem.relPath) // Mark as processed to avoid duplicates
            }
          }

          results.sort()
          return [results, true]
        }
      }
    }
  }

  results.sort()
  // If we exhaust the BFS queue, we did not reach maxCount
  return [results, false]
}
