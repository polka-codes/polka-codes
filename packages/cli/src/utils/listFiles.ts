import { promises as fs } from 'node:fs'
import { join, relative, resolve } from 'node:path'
import ignore, { type Ignore } from 'ignore'

/** Default patterns commonly ignored in projects of various languages. */
const DEFAULT_IGNORES = [
  '__pycache__',
  '.DS_Store',
  '.env',
  '.git',
  '.idea',
  '.svn',
  '.temp',
  '.vscode',
  'coverage',
  'dist',
  'node_modules',
  'out',
  'Thumbs.db',
]

/**
 * Reads a `.gitignore` file in `dirPath` (if it exists) and appends its lines
 * to the `basePatterns`. Returns a new array without mutating the original.
 */
async function extendPatterns(basePatterns: string[], dirPath: string): Promise<string[]> {
  try {
    const gitignorePath = join(dirPath, '.gitignore')
    const content = await fs.readFile(gitignorePath, 'utf8')
    const lines = content.split(/\r?\n/).filter(Boolean)
    return [...basePatterns, ...lines]
  } catch {
    // No .gitignore or unreadable
    return basePatterns
  }
}

/** Creates an `ignore` instance from the given patterns. */
function createIgnore(patterns: string[]): Ignore {
  return ignore().add(patterns)
}

/**
 * Lists files under `dirPath` in BFS order, respecting:
 *   - A default set of ignores
 *   - A root .gitignore under `cwd`
 *   - Any .gitignore files in child directories (merged as we go)
 *
 * Returns `[files, limitReached]`:
 *   - `files` is the array of file paths (relative to `cwd`)
 *   - `limitReached` is `true` if `maxCount` was hit, otherwise `false`
 */
export async function listFiles(dirPath: string, recursive: boolean, maxCount: number, cwd: string): Promise<[string[], boolean]> {
  // Merge default ignores with root .gitignore (if found)
  let rootPatterns = [...DEFAULT_IGNORES]
  try {
    const rootGitignore = await fs.readFile(join(cwd, '.gitignore'), 'utf8')
    const lines = rootGitignore.split(/\r?\n/).filter(Boolean)
    rootPatterns = [...rootPatterns, ...lines]
  } catch {
    // No .gitignore at root or unreadable; ignore silently
  }

  // Final results (relative to `cwd`) and indicator if we reached the limit
  const results: string[] = []

  // BFS queue
  // Each entry holds the directory path and the patterns inherited from its parent
  const queue: Array<{ path: string; patterns: string[] }> = [{ path: resolve(dirPath), patterns: rootPatterns }]

  // Perform BFS until queue is empty or maxCount is reached
  while (queue.length > 0) {
    // biome-ignore lint/style/noNonNullAssertion: checked above
    const { path: currentPath, patterns: parentPatterns } = queue.shift()!

    // Merge parent's patterns with local .gitignore
    const mergedPatterns = await extendPatterns(parentPatterns, currentPath)
    const folderIg = createIgnore(mergedPatterns)

    const entries = await fs.readdir(currentPath, { withFileTypes: true })

    for (const entry of entries) {
      const fullPath = join(currentPath, entry.name)
      // Convert full path to something relative to `cwd`
      const relPath = relative(cwd, fullPath).replace(/\\/g, '/')

      if (folderIg.ignores(relPath)) {
        continue // Skip ignored entries
      }

      if (entry.isDirectory()) {
        if (recursive) {
          queue.push({ path: fullPath, patterns: mergedPatterns })
        }
      } else {
        results.push(relPath)
        if (results.length >= maxCount) {
          results.sort()
          // Stop searching as soon as we reach maxCount
          return [results, true]
        }
      }
    }
  }

  results.sort()
  // If we exhaust the BFS queue, we did not reach maxCount
  return [results, false]
}
