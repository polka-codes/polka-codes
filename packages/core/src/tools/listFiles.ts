import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../tool'
import type { FilesystemProvider } from './provider'

export const toolInfo = {
  name: 'list_files',
  description:
    'Request to list files and directories within the specified directory. If recursive is true, it will list all files and directories recursively. If recursive is false or not provided, it will only list the top-level contents. Do not use this tool to confirm the existence of files you may have created, as the user will let you know if the files were created successfully or not.',
  parameters: z.object({
    path: z
      .string()
      .describe('The path of the directory to list contents for (relative to the current working directory)')
      .meta({ usageValue: 'Directory path here' }),
    maxCount: z.coerce
      .number()
      .optional()
      .default(2000)
      .describe('The maximum number of files to list. Default to 2000')
      .meta({ usageValue: 'Maximum number of files to list (optional)' }),
    recursive: z
      .preprocess((val) => {
        if (typeof val === 'string') {
          const lower = val.toLowerCase()
          if (lower === 'false') return false
          if (lower === 'true') return true
        }
        return val
      }, z.boolean().optional().default(true))
      .describe('Whether to list files recursively. Use true for recursive listing, false or omit for top-level only.')
      .meta({ usageValue: 'true or false (optional)' }),
  }),
  examples: [
    {
      description: 'Request to list files',
      parameters: [
        {
          name: 'path',
          value: 'src',
        },
        {
          name: 'maxCount',
          value: '100',
        },
      ],
    },
  ],
  permissionLevel: PermissionLevel.Read,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, FilesystemProvider> = async (provider, args) => {
  if (!provider.listFiles) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to list files. Abort.',
    }
  }

  const { path, maxCount, recursive } = toolInfo.parameters.parse(args)
  const [files, limitReached] = await provider.listFiles(path, recursive, maxCount)

  return {
    type: ToolResponseType.Reply,
    message: `<list_files_path>${path}</list_files_path>
<list_files_files>
${files.join('\n')}
</list_files_files>
<list_files_truncated>${limitReached}</list_files_truncated>`,
  }
}

export const isAvailable = (provider: FilesystemProvider): boolean => {
  return !!provider.listFiles
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
