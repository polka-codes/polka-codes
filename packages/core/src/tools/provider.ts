export type FilesystemProvider = {
  readFile?: (path: string) => Promise<string>
  writeFile?: (path: string, content: string) => Promise<void>
  removeFile?: (path: string) => Promise<void>
  listFiles?: (path: string, recursive: boolean, maxCount: number) => Promise<[string[], boolean]>
  searchFiles?: (path: string, regex: string, filePattern: string) => Promise<string[]>
  listCodeDefinitionNames?: (path: string) => Promise<string[]>
}

export type CommandProvider = {
  executeCommand?: (command: string, needApprove: boolean) => Promise<{ stdout: string; stderr: string; exitCode: number }>
}

export type InteractionProvider = {
  askFollowupQuestion?: (question: string, options: string[]) => Promise<string>
  attemptCompletion?: (result: string) => Promise<string | undefined>
}

export type ToolProvider = FilesystemProvider & CommandProvider & InteractionProvider

export class MockProvider implements ToolProvider {
  async readFile(path: string): Promise<string> {
    return 'mock content'
  }

  async writeFile(path: string, content: string): Promise<void> {
    return
  }

  async removeFile(path: string): Promise<void> {
    return
  }

  async listFiles(path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> {
    return [['mock-file.txt'], false]
  }

  async searchFiles(path: string, regex: string, filePattern: string): Promise<string[]> {
    return ['mock-file.txt']
  }

  async listCodeDefinitionNames(path: string): Promise<string[]> {
    return ['mockDefinition']
  }

  async executeCommand(command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> {
    return { stdout: 'mock output', stderr: '', exitCode: 0 }
  }

  async askFollowupQuestion(question: string, options?: string[]): Promise<string> {
    return 'mock answer'
  }

  async attemptCompletion(result: string): Promise<string | undefined> {
    return 'mock completion'
  }
}
