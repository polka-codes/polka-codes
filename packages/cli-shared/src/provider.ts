import { spawn } from 'node:child_process'
import { mkdir, readFile, rename, unlink, writeFile } from 'node:fs/promises'
import { dirname, normalize, resolve } from 'node:path'
import { input, select } from '@inquirer/prompts'
import type { ToolProvider } from '@polka-codes/core'
import ignore from 'ignore'
import { lookup } from 'mime-types'
import { checkRipgrep } from './utils/checkRipgrep'
import { listFiles } from './utils/listFiles'
import { searchFiles } from './utils/searchFiles'

export type ProviderOptions = {
  command?: {
    onStarted(command: string): void
    onStdout(data: string): void
    onStderr(data: string): void
    onExit(code: number): void
    onError(error: unknown): void
  }
  excludeFiles?: string[]
  summarizeOutput?: (stdout: string, stderr: string) => Promise<string | undefined>
  summaryThreshold?: number
}

export const getProvider = (options: ProviderOptions = {}): ToolProvider => {
  const ig = ignore().add(options.excludeFiles ?? [])
  const memoryStore: Record<string, string> = {}
  const provider: ToolProvider = {
    listTopics: async (): Promise<string[]> => {
      return Object.keys(memoryStore)
    },
    read: async (topic: string): Promise<string | undefined> => {
      return memoryStore[topic]
    },
    append: async (topic: string, content: string): Promise<void> => {
      if (memoryStore[topic]) {
        memoryStore[topic] += content
      } else {
        memoryStore[topic] = content
      }
    },
    replace: async (topic: string, content: string): Promise<void> => {
      memoryStore[topic] = content
    },
    remove: async (topic: string): Promise<void> => {
      delete memoryStore[topic]
    },
    readFile: async (path: string, includeIgnored: boolean): Promise<string | undefined> => {
      if (!includeIgnored && ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      try {
        return await readFile(path, 'utf8')
      } catch (_e) {
        return undefined
      }
    },
    writeFile: async (path: string, content: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      // generate parent directories if they don't exist
      await mkdir(dirname(path), { recursive: true })

      return await writeFile(path, content, 'utf8')
    },
    removeFile: async (path: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      return await unlink(path)
    },
    renameFile: async (sourcePath: string, targetPath: string): Promise<void> => {
      if (ig.ignores(sourcePath) || ig.ignores(targetPath)) {
        throw new Error(`Not allow to access file ${sourcePath} or ${targetPath}`)
      }
      return await rename(sourcePath, targetPath)
    },
    listFiles: async (path: string, recursive: boolean, maxCount: number, includeIgnored: boolean): Promise<[string[], boolean]> => {
      return await listFiles(path, recursive, maxCount, process.cwd(), options.excludeFiles, includeIgnored)
    },
    readBinaryFile: async (url: string) => {
      if (url.startsWith('file://')) {
        const filePath = decodeURIComponent(url.substring('file://'.length))
        const resolvedPath = normalize(resolve(process.cwd(), filePath))

        if (!resolvedPath.startsWith(process.cwd())) {
          throw new Error(`Access to file path "${filePath}" is restricted.`)
        }

        const data = await readFile(resolvedPath)
        const mediaType = lookup(resolvedPath) || 'application/octet-stream'

        return {
          base64Data: data.toString('base64'),
          mediaType,
        }
      }
      const response = await fetch(url)
      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }
      const data = await response.arrayBuffer()
      const mediaType = lookup(url) || 'application/octet-stream'

      return {
        base64Data: Buffer.from(data).toString('base64'),
        mediaType,
      }
    },

    executeCommand: (
      command: string,
      _needApprove: boolean,
    ): Promise<{ stdout: string; stderr: string; exitCode: number; summary?: string }> => {
      // TODO: add timeout

      return new Promise((resolve, reject) => {
        // spawn a shell to execute the command

        options.command?.onStarted(command)

        const child = spawn(command, [], {
          shell: true,
          stdio: 'pipe',
        })

        let stdoutText = ''
        let stderrText = ''

        child.stdout.on('data', (data) => {
          const dataStr = data.toString()
          options.command?.onStdout(dataStr)
          stdoutText += dataStr
        })

        child.stderr.on('data', (data) => {
          const dataStr = data.toString()
          options.command?.onStderr(dataStr)
          stderrText += dataStr
        })

        child.on('close', async (code) => {
          options.command?.onExit(code ?? 0)
          const totalLength = stdoutText.length + stderrText.length
          if (totalLength > (options.summaryThreshold ?? 5000) && options.summarizeOutput) {
            try {
              const summary = await options.summarizeOutput(stdoutText, stderrText)
              if (summary) {
                resolve({
                  summary,
                  stdout: stdoutText,
                  stderr: stderrText,
                  exitCode: code ?? 0,
                })
                return
              }
            } catch (_e) {
              console.error('Summarization failed:', _e)
            }
          }
          resolve({
            stdout: stdoutText,
            stderr: stderrText,
            exitCode: code ?? 0,
          })
        })

        child.on('error', (err) => {
          options.command?.onError(err)
          reject(err)
        })
      })
    },
    askFollowupQuestion: async (question: string, answerOptions: string[]): Promise<string> => {
      if (answerOptions.length === 0) {
        return await input({ message: question })
      }

      const otherMessage = 'Other (enter text)'
      answerOptions.push(otherMessage)
      const answer = await select({
        message: question,
        choices: answerOptions.map((option) => ({ name: option, value: option })),
      })

      if (answer === otherMessage) {
        return await input({ message: 'Enter your answer:' })
      }
      return answer
    },
    fetchUrl: async (url: string): Promise<string> => {
      const isRaw = url.startsWith('https://raw.githubusercontent.com/')

      const urlToFetch = isRaw ? url : `https://r.jina.ai/${url}`

      try {
        const response = await fetch(urlToFetch)
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`)
        }
        return await response.text()
      } catch (error) {
        console.error('Error fetching URL:', error)
        throw error
      }
    },
  }

  if (checkRipgrep()) {
    provider.searchFiles = async (path: string, regex: string, filePattern: string): Promise<string[]> => {
      return await searchFiles(path, regex, filePattern, process.cwd(), options.excludeFiles)
    }
  } else {
    console.error(
      'Error: ripgrep (rg) is not installed. Search file tool is disabled. Please install it from https://github.com/BurntSushi/ripgrep#installation',
    )
  }

  return provider
}
