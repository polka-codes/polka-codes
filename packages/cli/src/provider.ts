import { spawn } from 'node:child_process'
import { mkdir, readFile, writeFile } from 'node:fs/promises'
import { dirname } from 'node:path'
import ignore from 'ignore'

import type { ToolProvider } from '@polka-codes/core'
import { listFiles } from './utils/listFiles'
import { searchFiles } from './utils/searchFiles'

export type ProviderOptions = {
  command: {
    onStarted(command: string): void
    onStdout(data: string): void
    onStderr(data: string): void
    onExit(code: number): void
    onError(error: unknown): void
  }
  excludeFiles?: string[]
}

export const getProvider = (options: ProviderOptions): ToolProvider => {
  const ig = ignore().add(options.excludeFiles ?? [])
  return {
    readFile: async (path: string): Promise<string> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      return await readFile(path, 'utf8')
    },
    writeFile: async (path: string, content: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      // generate parent directories if they don't exist
      await mkdir(dirname(path), { recursive: true })

      return await writeFile(path, content, 'utf8')
    },
    listFiles: async (path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> => {
      return await listFiles(path, recursive, maxCount, dirname(path), options.excludeFiles)
    },
    searchFiles: async (path: string, regex: string, filePattern: string): Promise<string[]> => {
      return await searchFiles(path, regex, filePattern, dirname(path), options.excludeFiles)
    },
    // listCodeDefinitionNames: async (path: string) => Promise<string[]> {},

    executeCommand: (command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> => {
      // TODO: add timeout

      return new Promise((resolve, reject) => {
        // spawn a shell to execute the command

        options.command.onStarted(command)

        const child = spawn(command, [], {
          shell: true,
          stdio: 'pipe',
        })

        let stdoutText = ''
        let stderrText = ''

        child.stdout.on('data', (data) => {
          const dataStr = data.toString()
          options.command.onStdout(dataStr)
          stdoutText += dataStr
        })

        child.stderr.on('data', (data) => {
          const dataStr = data.toString()
          options.command.onStderr(dataStr)
          stderrText += dataStr
        })

        child.on('close', (code) => {
          options.command.onExit(code ?? 0)
          resolve({
            stdout: stdoutText,
            stderr: stderrText,
            exitCode: code ?? 0,
          })
        })

        child.on('error', (err) => {
          options.command.onError(err)
          reject(err)
        })
      })
    },
    // askFollowupQuestion: async (question: string, options: string[]) => Promise<string> {},
    // attemptCompletion: async (result: string) => Promise<string | undefined> {},
  }
}
