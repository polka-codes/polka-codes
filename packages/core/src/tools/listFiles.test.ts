import { describe, expect, it, spyOn } from 'bun:test'
import listFiles from './listFiles'
import { MockProvider } from './provider'

describe('listFiles', () => {
  it('should return file list', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'listFiles').mockResolvedValue([['file1.ts', 'file2.ts'], false])

    const result = await listFiles.handler(mockProvider, {
      path: 'src',
      maxCount: '10',
      includeIgnored: 'false',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.listFiles).toHaveBeenCalledWith('src', true, 10, false)
  })

  it('should handle non recursive listing', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'listFiles').mockResolvedValue([['file1.ts', 'file2.ts'], false])

    const result = await listFiles.handler(mockProvider, {
      path: 'src',
      recursive: 'false',
      includeIgnored: 'false',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.listFiles).toHaveBeenCalledWith('src', false, 2000, false)
  })

  it('should handle empty directory', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'listFiles').mockResolvedValue([[], true])

    const result = await listFiles.handler(mockProvider, {
      path: 'empty-dir',
      includeIgnored: 'false',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.listFiles).toHaveBeenCalledWith('empty-dir', true, 2000, false)
  })

  it('should handle max count with truncation', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'listFiles').mockResolvedValue([['file1.ts', 'file2.ts'], true])

    const result = await listFiles.handler(mockProvider, {
      path: 'src',
      maxCount: '1',
      includeIgnored: 'false',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.listFiles).toHaveBeenCalledWith('src', true, 1, false)
  })

  it('should handle errors', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'listFiles').mockRejectedValue(new Error('Directory not found'))

    const result = listFiles.handler(mockProvider, {
      path: 'invalid-path',
      includeIgnored: 'false',
    })

    await expect(result).rejects.toMatchSnapshot()
    expect(mockProvider.listFiles).toHaveBeenCalledWith('invalid-path', true, 2000, false)
  })
})
