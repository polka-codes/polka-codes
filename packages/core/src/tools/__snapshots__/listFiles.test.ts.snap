// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`listFiles should return file list 1`] = `
{
  "message": {
    "type": "text",
    "value": 
"<list_files_path>src</list_files_path>
<list_files_files>
file1.ts
file2.ts
</list_files_files>
<list_files_truncated>false</list_files_truncated>"
,
  },
  "type": "Reply",
}
`;

exports[`listFiles should handle non recursive listing 1`] = `
{
  "message": {
    "type": "text",
    "value": 
"<list_files_path>src</list_files_path>
<list_files_files>
file1.ts
file2.ts
</list_files_files>
<list_files_truncated>false</list_files_truncated>"
,
  },
  "type": "Reply",
}
`;

exports[`listFiles should handle empty directory 1`] = `
{
  "message": {
    "type": "text",
    "value": 
"<list_files_path>empty-dir</list_files_path>
<list_files_files>

</list_files_files>
<list_files_truncated>true</list_files_truncated>"
,
  },
  "type": "Reply",
}
`;

exports[`listFiles should handle max count with truncation 1`] = `
{
  "message": {
    "type": "text",
    "value": 
"<list_files_path>src</list_files_path>
<list_files_files>
file1.ts
file2.ts
</list_files_files>
<list_files_truncated>true</list_files_truncated>"
,
  },
  "type": "Reply",
}
`;

exports[`listFiles should handle errors 1`] = `[Error: Directory not found]`;
